File: 42_Trapping_Rain_Water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        if(height.size() <= 0) return 0; 
        
        vector<int> left;
        int currentMax = height[0]; 
        for(int i = 0; i < height.size(); i++){
            currentMax = max(currentMax, height[i]);
            left.push_back(currentMax); 
        }
        currentMax = height[height.size()-1];
        vector<int> right; 
        for(int j = height.size() - 1; j >= 0; j--){
            currentMax = std::max(currentMax, height[j]); 
            right.push_back(currentMax); 
        }
        
        std::reverse(right.begin(), right.end());
        
        int water = 0; 
        for(int i = 0; i < height.size(); i++){
            int minHeight = std::min(left[i], right[i]); 
            if(height[i] < minHeight){
                water += (minHeight - height[i]); 
            }
        }
        return water;
    }
};

File: 42_Trapping_Rain_Water.py
'''
Pseudo Code:
    Create 1 array (iterate left to right) 
    Calculate left max 
    Create 1 array (iterate right to left)
    Calculate right max 
    Create minHeight array (min(leftMax, rightMax)) 
    Iterate through Height:
        if height < minHeight: 
            water += minHeight - height 
        else:
            water = 0 

Time Complexity O(n) || O(n) space complexity 

'''

class Solution:
    def trap(self, height: List[int]) -> int:
        if len(height) <= 0:
            return 0 
        
        left = []
        leftMax = height[0]
        for i in height: 
            leftMax = max(leftMax, i) 
            left.append(leftMax) 

        right = []
        rightMax = height[-1] 
        for j in height[::-1]:
            rightMax = max(rightMax, j) 
            right.append(rightMax) 
        right.reverse()
        minHeight = [] 
    
        for i in range(0, len(left)):
            minHeight.append( min(left[i], right[i])) 
        c = 0 
        water = 0 
        for k in height:
            if k < minHeight[c]: 
                water += minHeight[c] - k 
            else: 
                water += 0 
            c += 1 
        return water
        